File: jumpGame1DFS.java
// Time Complexity : O(exp)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
import java.util.*;

class jump1DFS {
    Set<Integer> visited;
    public boolean canJump(int[] nums) {
        if(nums.length == 0 || nums == null)
            return false;
        
        if(nums.length == 1)
            return true;
        
        visited = new HashSet<>();
        return dfs(0, nums);
        
    }
    
    private boolean dfs(int index, int[] nums){
        //base
        if(visited.contains(index)) return false;
        
        if(index == nums.length - 1) return true;
        
        //logic
        visited.add(index);
        for(int i = 1; i <= nums[index]; i++){
            if(dfs(index + i, nums)) return true;
        }
        
        return false;
    }
}

File: jump_game_1_dfs.py
# Time Complexity : O(N^M) where N elements in the array and M is max (jump capacity/strength) number in the array
# Space Complexity : O(N) where N is number of elements in the array.
# Did this code successfully run on Leetcode : No (Time Limit Exceeded)
# Any problem you faced while coding this : No

#DFS Solution
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        if len(nums) == 1: return True
        hashset = set()
        
        return self.dfs(nums, 0, hashset)
        
    def dfs(self, nums, index, hashset):
        #base
        if index >= len(nums) - 1:
            return True
        
        if index in hashset:
            return False
        
        #logic
        hashset.add(index)
        for i in range(1, nums[index] + 1):
            nextIndex = index + i 
            if self.dfs(nums, nextIndex, hashset) : return True
                    
        return False